source/common/buffer/watermark_buffer.h

```diff
@@ -72,15 +74,98 @@ class WatermarkBuffer : public OwnedImpl {
 
 using WatermarkBufferPtr = std::unique_ptr<WatermarkBuffer>;
 
+class WatermarkBufferFactory;
+
+/**
+ * A BufferMemoryAccountImpl tracks allocated bytes across associated buffers and
+ * slices that originate from those buffers, or are untagged and pass through an
+ * associated buffer.
+ *
+ * This BufferMemoryAccount is produced by the *WatermarkBufferFactory*.
+ */
+class BufferMemoryAccountImpl : public BufferMemoryAccount {
+public:
+  // Used to create the account, and complete wiring with the factory
+  // and shared_this_.
+  static BufferMemoryAccountSharedPtr createAccount(WatermarkBufferFactory* factory,
+                                                    Http::StreamResetHandler* reset_handler);
+  ~BufferMemoryAccountImpl() override { ASSERT(buffer_memory_allocated_ == 0); }
+
+  // Make not copyable
+  BufferMemoryAccountImpl(const BufferMemoryAccountImpl&) = delete;
+  BufferMemoryAccountImpl& operator=(const BufferMemoryAccountImpl&) = delete;
+
+  // Make not movable.
+  BufferMemoryAccountImpl(BufferMemoryAccountImpl&&) = delete;
+  BufferMemoryAccountImpl& operator=(BufferMemoryAccountImpl&&) = delete;
+
+  uint64_t balance() const override { return buffer_memory_allocated_; }
+  void charge(uint64_t amount) override;
+  void credit(uint64_t amount) override;
+
+  void clearDownstream() override;
+
+  void resetDownstream(Http::StreamResetReason reason) override {
+    if (reset_handler_ != nullptr) {
+      reset_handler_->resetStream(reason);
```

Comment:
You may be making an assumption that `reset_handler_->resetStream(reason);` calls `clearDownstream()`

You may run into some problems in the watermark factory while iterating through accounts and calling `resetDownstream` if some implementation doesn't invoke `clearDownstream()` from `resetStream`.  I'm fairly sure that some of the stream implementations like HTTP2 do not call `clearDownstream()` from `resetStream` since they use the deferred delete mechanism provided by the dispatcher.  Another potential danger is you calling `reset_handler_->resetStream();` on a stream that is waiting to be deleted via deferred deletion.  If the stream doesn't correctly handle the case by ignoring the resetStream we may trigger a crash.

Reply:
Done.